src/transformers/models/speecht5/modeling_speecht5.py

```diff
@@ -2606,36 +2617,69 @@ def _generate_speech(
+        last_decoder_output = decoder_out.last_hidden_state.squeeze(1)
         past_key_values = decoder_out.past_key_values
 
         # Predict the new mel spectrum for this step in the sequence.
         spectrum = model.speech_decoder_postnet.feat_out(last_decoder_output)
-        spectrum = spectrum.view(model.config.reduction_factor, model.config.num_mel_bins)
+        spectrum = spectrum.view(bsz, model.config.reduction_factor, model.config.num_mel_bins)
         spectrogram.append(spectrum)
 
         # Extend the output sequence with the new mel spectrum.
-        output_sequence = torch.cat((output_sequence, spectrum[-1].view(1, 1, model.config.num_mel_bins)), dim=1)
-
+        new_spectrogram = spectrum[:, -1, :].view(bsz, 1, model.config.num_mel_bins)
+        output_sequence = torch.cat((output_sequence, new_spectrogram), dim=1)
         # Predict the probability that this is the stop token.
         prob = torch.sigmoid(model.speech_decoder_postnet.prob_out(last_decoder_output))
 
-        # Finished when stop token or maximum length is reached.
-        if idx >= minlen and (int(sum(prob >= threshold)) > 0 or idx >= maxlen):
-            spectrogram = torch.cat(spectrogram, dim=0).unsqueeze(0)
-            spectrogram = model.speech_decoder_postnet.postnet(spectrogram)
-            spectrogram = spectrogram.squeeze(0)
-            break
-
-    if vocoder is not None:
-        outputs = vocoder(spectrogram)
+        if idx < minlen:
+            continue
+        else:
+            # If the generation loop is less than maximum length time, check the ones in the batch that have met
+            # the prob threshold. Otherwise, assume all have met thresholds and fill other spectrograms for the batch.
+            if idx < maxlen:
+                meet_thresholds = torch.sum(prob, dim=-1) >= threshold
+                meet_indexes = torch.where(meet_thresholds)[0].tolist()
+            else:
+                meet_indexes = range(len(prob))
+            meet_indexes = [i for i in meet_indexes if i not in result_spectrogram]
+            if len(meet_indexes) > 0:
+                spectrograms = torch.stack(spectrogram)
+                spectrograms = spectrograms.transpose(0, 1).flatten(1, 2)
+                spectrograms = model.speech_decoder_postnet.postnet(spectrograms)
+                for meet_index in meet_indexes:
+                    result_spectrogram[meet_index] = spectrograms[meet_index]
+            if len(result_spectrogram) >= bsz:
+                break
+    spectrograms = [result_spectrogram[i] for i in range(len(result_spectrogram))]
+    if not return_concrete_lengths:
+        spectrogram = spectrograms[0]
+        if vocoder is not None:
+            outputs = vocoder(spectrogram)
+        else:
+            outputs = spectrogram
+        if output_cross_attentions:
+            cross_attentions = torch.cat(cross_attentions, dim=2)
+            outputs = (outputs, cross_attentions)
```

Comment:
So if `return_concrete_lengths=False`, it only return the first sample of the batch ? It probably should have the same behavior than with `return_concrete_lengths=True` except that it doesn't return output length!

Reply:
Oh so what you mean is also considering `return_concrete_lengths=False` when feeding batched inputs? Under what circumstance do you think users do not need the concrete lengths to filter out the concrete part of the batched outputs? IMO, if they plan to feed batched inputs, they should always set `return_concrete_lengths=True`.

Reply:
In the latest commit I have updated the code based on your request, and also added a test. Spectrograms/Waveform results are the same with return_concrete_lengths=True/False.

Reply:
However, when batch_size==1, 

1)  return_concrete_lengths=False
to make it compatible to old functionality, I will return the result without **batch_size** dimension

2) return_concerte_lengths=True
to make it easy to use for users who need batch input but not aware of the batch_size, I will return the result with the **batch_size** dimension, namely shape (1, ...)

Reply:
Nice, thanks for the explanation and for addressing my comments, LGTM !


a qq for my curiosity:

when outputting cross-attentions, you make [here](https://github.com/huggingface/transformers/blob/4cef843994276e594e0a0a4f15f56f2d91b05c5d/src/transformers/models/speecht5/modeling_speecht5.py#L2677-L2682) an operation that you don't make [here](https://github.com/huggingface/transformers/blob/4cef843994276e594e0a0a4f15f56f2d91b05c5d/src/transformers/models/speecht5/modeling_speecht5.py#L2661-L2662), is there a difference here ?

Reply:
Hi @ylacombe , yes, there is a difference. The `cross_attentions` after line 2618 should be a list of shape e.g. (bsz*6, ..., 1, ...), and considering we are doing batch generation and we need batch_size as our first dimension, we need to split the bsz*6 as two dimension (bsz, 6, ...), and that is why I wrote

```
cross_attentions = cross_attentions.view(
                bsz, int(cross_attentions.size(0) / bsz), *cross_attentions.size()[-3:]
            )
```

By the way, in case of the consistency, I should also modify the corresponding logics of the case 2 (among the 4 cases) I mentioned above, and I will include it in the latest commit. Thanks:)